cf-serverd/server_transform.c

```diff
@@ -543,10 +542,37 @@ static void KeepControlPromises(EvalContext *ctx, const Policy *policy, GenericA
 
 /*********************************************************************/
 
-static void KeepContextBundles(EvalContext *ctx, const Policy *policy)
+/* Sequence in which server promise types should be evaluated */
+static const char *const SERVER_TYPESEQUENCE[] =
+{
+    "vars",
+    "classes",
+    "roles",
+    "access",
+    NULL
+};
+
+/* Check if promise is NOT belonging to default server types 
+ * (see SERVER_TYPESEQUENCE)*/
+static bool IsPromiseTypeNotInServerTypeSequence(const char *promise_type)
+{
+    for (int type = 0; SERVER_TYPESEQUENCE[type] != NULL; type++)
+    {
+        if (strncmp(promise_type, SERVER_TYPESEQUENCE[type], 
```

Comment:
On Thu, Feb 19, 2015 at 4:24 PM, pasinskim <notifications@github.com> wrote:

> In cf-serverd/server_transform.c
> <https://github.com/cfengine/core/pull/2111#discussion_r24995120>:
>
> > +static const char *const SERVER_TYPESEQUENCE[] =
> > +{
> > +    "vars",
> > +    "classes",
> > +    "roles",
> > +    "access",
> > +    NULL
> > +};
> > +
> > +/* Check if promise is NOT belonging to default server types
> > + * (see SERVER_TYPESEQUENCE)*/
> > +static bool IsPromiseTypeNotInServerTypeSequence(const char *promise_type)
> > +{
> > +    for (int type = 0; SERVER_TYPESEQUENCE[type] != NULL; type++)
> > +    {
> > +        if (strncmp(promise_type, SERVER_TYPESEQUENCE[type],
>
> I think no specific reason. Can change to strcmp as both strings are null
> terminated for sure.
>
Yes, because `strncmp()` has a different purpose: it compares to true if
one string is **prefix** of another. E.g.
```strncmp("123", "12345", 3);```